url27a7ae3508_serkan_ozal_ocean_of_memories_tgz-pJ8-com_zeroturnaround_rebellabs_oceanofmemories_article1_objectlayout_ObjectMemoryLayoutDemoJ8/src/com/zeroturnaround/rebellabs/oceanofmemories/common/util/JvmUtil.java

```diff
@@ -0,0 +1,2125 @@
+        return b.toString();
+    }
+
+    public static long alignObjectSize(long size) {
+        size += (long) options.getObjectAlignment() - 1L;
+        return size - (size % options.getObjectAlignment());
+    }
+
+    public static long sizeOf(Object obj) {
+        ArrayList<Object> stack = new ArrayList<Object>();
+        stack.add(obj);
+        return measureSizeOf(stack);
+    }
+
+    public static long sizeOfAll(Object... objects) {
+        return sizeOfAll(Arrays.asList(objects));
+    }
+
+    public static long sizeOfAll(Iterable<Object> objects) {
+        final ArrayList<Object> stack;
+        if (objects instanceof Collection<?>) {
+            stack = new ArrayList<Object>(((Collection<?>) objects).size());
+        } else {
+            stack = new ArrayList<Object>();
+        }
+
+        for (Object o : objects) {
+            stack.add(o);
+        }
+
+        return measureSizeOf(stack);
+    }
+
+    public static long shallowSizeOf(Object obj) {
+        if (obj == null) {
+            return 0;
+        }
+        final Class<?> clz = obj.getClass();
+        if (clz.isArray()) {
+            return shallowSizeOfArray(obj);
+        } else {
+            return sizeOf(clz);
+        }
+    }
+
+    public static long sizeOf(Class<?> clazz) {
+        return getClassInfo(clazz).size;
+    }
+
+    public static long shallowSizeOfAll(Object... objects) {
+        return shallowSizeOfAll(Arrays.asList(objects));
+    }
+
+    public static long shallowSizeOfAll(Iterable<Object> objects) {
+        long sum = 0;
+        for (Object o : objects) {
+            sum += shallowSizeOf(o);
+        }
+        return sum;
+    }
+
+    private static long shallowSizeOfInstance(Class<?> clazz) {
+        if (clazz.isArray()) {
+            throw new IllegalArgumentException(
+                    "This method does not work with array classes.");
+        }
+        if (clazz.isPrimitive()) {
+            return sizeOfType(clazz);
+        }
+        long size = headerSize;
+
+        for (; clazz != null; clazz = clazz.getSuperclass()) {
+            final Field[] fields = clazz.getDeclaredFields();
+            for (Field f : fields) {
+                if (!Modifier.isStatic(f.getModifiers())) {
+                    size = adjustForField(size, f);
+                }
+            }
+        }
+        return alignObjectSize(size);
+    }
+
+    private static long shallowSizeOfArray(Object array) {
+        long size = arrayHeaderSize;
+        final int len = Array.getLength(array);
+        if (len > 0) {
+            Class<?> arrayElementClazz = array.getClass().getComponentType();
+            if (arrayElementClazz.isPrimitive()) {
+                size += (long) len * sizeOfType(arrayElementClazz);
+            } else {
+                size += (long) options.referenceSize * len;
+            }
+        }
+        return alignObjectSize(size);
+    }
+
+    private static long measureSizeOf(ArrayList<Object> stack) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.